lib/helper/GraphQLDataFactory.js

```diff
@@ -0,0 +1,415 @@
+ * Uses data generators like [rosie](https://github.com/rosiejs/rosie) or factory girl to create new record.
+ *
+ * By defining a factory you set the rules of how data is generated.
+ * This data will be saved on server via GraphQL queries and deleted in the end of a test.
+ *
+ * ## Use Case
+ *
+ * Acceptance tests interact with a websites using UI and real browser.
+ * There is no way to create data for a specific test other than from user interface.
+ * That makes tests slow and fragile. Instead of testing a single feature you need to follow all creation/removal process.
+ *
+ * This helper solves this problem.
+ * If a web application has GraphQL support, it can be used to create and delete test records.
+ * By combining GraphQL with Factories you can easily create records for tests:
+ *
+ * ```js
+ * I.have('user', { login: 'davert', email: 'davert@mail.com' });
+ * let id = await I.have('post', { title: 'My first post'});
+ * I.haveMultiple('comment', 3, {post_id: id});
+ * ```
+ *
+ * To make this work you need
+ *
+ * 1. GraphQL endpoint which allows to perform create / delete requests and
+ * 2. define data generation rules
+ *
+ * ### Setup
+ *
+ * Install [Rosie](https://github.com/rosiejs/rosie) and [Faker](https://www.npmjs.com/package/faker) libraries.
+ *
+ * ```sh
+ * npm i rosie faker --save-dev
+ * ```
+ *
+ * Create a factory file for a resource.
+ *
+ * See the example for Posts factories:
+ *
+ * ```js
+ * // tests/factories/posts.js
+ *
+ * var Factory = require('rosie').Factory;
+ * var faker = require('faker');
+ *
+ * module.exports = new Factory()
+ *    // no need to set id, it will be set by GraphQL resolver
+ *    .attr('author', () => faker.name.findName())
+ *    .attr('title', () => faker.lorem.sentence())
+ *    .attr('body', () => faker.lorem.paragraph());
+ * ```
+ * For more options see [rosie documentation](https://github.com/rosiejs/rosie).
+ *
+ * Then configure GraphQLDataHelper to match factories and GraphQL schema:
+
+ * ### Configuration
+ *
+ * GraphQLDataFactory has following config options:
+ *
+ * * `endpoint`: base URL for the GraphQL server.
+ * * `cleanup` (default: true): should inserted records be deleted up after tests
+ * * `factories`: list of defined factories
+ * * `returnId` (default: false): return id instead of a complete response when creating items.
+ * * `headers`: list of headers
+ * * `GraphQL`: configuration for GraphQL requests.
+ *
+ *
+ * See the example:
+ *
+ * ```js
+ *  GraphQLDataFactory: {
+ *    endpoint: "http://user.com/graphql",
+ *    cleanup: true,
+ *    headers: {
+ *      'Content-Type': 'application/json',
+ *      'Accept': 'application/json',
+ *    },
+ *    factories: {
+ *      user: {
+ *        factory: "./factories/post",
+ *        operations: {
+ *          create: {
+ *            query: 'mutation createUser($input: UserInput!) { createUser(input: $input) { id name }}',
+ *            operationName: 'createUser',
```

Comment:
I was thinking of this format for configuration:

```js
{ createUser: {
  query: `
  mutation createUser($input: UserInput!) {
    createUser(input: $input) {
      id
      name
      email
    }
  }
  `,
  factory: 'user',
  revert: (id) => {
    return {
      query: `mutation deleteUser($id: ID!) { deleteUser(id: $id) }`,
      variables: { id }
    }
  }
}
}
```

Then we can create objects like this:

```js
I.mutate('createUser');
```
when test ends, a `revert` mutation will be executed

Reply:
Ok, even better idea, for revert function to use the full response of initial data, this way we won't need `_fetchId` function at all

```js
  revert: (data) => {
    return {
      query: `mutation deleteUser($id: ID!) { deleteUser(id: $id) }`,
      variables: { id: data.id }
    }
  }
```

Reply:
Used data instead of id in revert.